File: LRU Cache.cpp
class LRUCache {
private:
    list<pair<int, int>> keyValueList;
    unordered_map<int, list<pair<int, int>>::iterator> pos;
    int capacity;
public:
    LRUCache(int _capacity) {
        capacity = _capacity;
    }
    
    int get(int key) {
        if(pos.count(key)) {
            // Reorder
            keyValueList.splice(keyValueList.begin(), keyValueList, pos[key]);
            return pos[key]->second;
        }
        return -1;
    }
    
    void put(int key, int value) {
        if(pos.count(key)) {
            // Update value of the key from the list
            pos[key]->second = value;
            // Reorder
            keyValueList.splice(keyValueList.begin(), keyValueList, pos[key]);
            return;
        }
        if(keyValueList.size() == capacity) {
            // Remove key
            int keyToBeDeleted = keyValueList.back().first;
            keyValueList.pop_back();
            pos.erase(keyToBeDeleted);
        }
        // Add key
        keyValueList.push_front({key, value});
        pos[key] = keyValueList.begin();
    }
};

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache* obj = new LRUCache(capacity);
 * int param_1 = obj->get(key);
 * obj->put(key,value);
 */

File: LRU Cache.py
class Node:
    def __init__(self, key, value):
        self.key = key
        self.val = value
        self.prev = None
        self.next = None

class LRUCache:

    def __init__(self, capacity: int):
        self.capacity = capacity
        self.cache = {}

        self.head = Node(0, 0)
        self.tail = Node(0, 0)
        self.head.next = self.tail
        self.tail.prev = self.head

    def _insert(self, node):
        node.next = self.head.next
        self.head.next.prev = node
        self.head.next = node
        node.prev = self.head
        
    
    def _remove(self, node):
        node.prev.next = node.next
        node.next.prev = node.prev

    def get(self, key: int) -> int:
        if key in self.cache:
            node = self.cache[key]
            self._remove(node)
            self._insert(node)
            return node.val
        
        return -1

    def put(self, key: int, value: int) -> None:
        # If key is there in cache always remove it
        if key in self.cache:
            node = self.cache[key]
            self._remove(node)

        node = Node(key, value)
        self._insert(node)
        self.cache[key] = node

        if len(self.cache) > self.capacity:
            node = self.tail.prev
            self._remove(node)
            del self.cache[node.key]



# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)
